src/awkward1/operations/structure.py

```diff
@@ -996,15 +990,37 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
             + awkward1._util.exception_suffix(__file__)
         )
 
-    batch = [contents[0]]
-    for x in contents[1:]:
-        if batch[-1].mergeable(x, mergebool=mergebool):
-            batch.append(x)
-        else:
-            collapsed = batch[0].mergemany(batch[1:])
-            batch = [collapsed.merge_as_union(x)]
+    posaxis = contents[0].axis_wrap_if_negative(axis)
+    if posaxis == 0:
+        batch = [contents[0]]
+        for x in contents[1:]:
+            if batch[-1].mergeable(x, mergebool=mergebool):
+                batch.append(x)
+            else:
+                collapsed = batch[0].mergemany(batch[1:])
+                batch = [collapsed.merge_as_union(x)]
+
+        out = batch[0].mergemany(batch[1:])
+
+    else:
+        allow_records=True
+        for x in contents[1:]:
+            if x.axis_wrap_if_negative(axis) != posaxis:
+                raise ValueError(
+                    "cannot concatenate arrays in different axis"
+                    + awkward1._util.exception_suffix(__file__)
+                )
+
+        def getfunction(inputs, depth):
+            if depth == posaxis:
+                out = inputs[0].mergemany_as_union(inputs[1:], 1)
+                return lambda: (out,)
+            else:
+                return None
+
+        out = awkward1._util.broadcast_and_apply(contents, getfunction,
+            behavior=awkward1._util.behaviorof(*arrays), allow_records=True)[0]
```

Comment:
This is the call to `broadcast_and_apply` that can take `allow_records=True` as an optional argument. Oh, you've already got it here...

Reply:
Here is where it fails:
```python
>       assert ak.concatenate([numbers[0], records[0]], 1) == [
            [1.1, 2.2, 3.3, {"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
            [],
            [4.4, 5.5, {"x": 2.2, "y": [1, 2]}],
            [6.6, 7.7, 8.8, 9.9, {"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
        ]

tests\test_0184-concatenate-operation.py:194: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\numpy\lib\mixins.py:25: in func
    return ufunc(self, other)
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\highlevel.py:1369: in __array_ufunc__
    return awkward1._connect._numpy.array_ufunc(ufunc, method, inputs, kwargs)
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_connect\_numpy.py:215: in array_ufunc
    inputs, getfunction, behavior, allow_records=False
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_util.py:996: in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_util.py:759: in apply
    outcontent = apply(nextinputs, depth + 1)
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_util.py:847: in apply
    outcontent = apply(nextinputs, depth + 1)
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_util.py:659: in apply
    outcontents.append(apply(nextinputs, depth))
C:\hostedtoolcache\windows\Python\2.7.18\x86\lib\site-packages\awkward1\_util.py:886: in apply
    deprecate(exception, "1.0.0", "2020-12-01")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
```

Reply:
Don't set `allow_records=True` in `_numpy.py:215` (the `array_ufunc` method). It was for that exact case that `allow_records` was introduced, to set it to `False`.

You are correctly setting `allow_records=True` in `structure.py:1021`.

The reason it's a parameter is so that functions like `concatenate` can set it to True while `array_ufunc` can set it to False. NumPy's ufuncs are not allowed to descend through RecordArrays unless a customization is defined, so that's why it was set to False in `array_ufunc`, and that's what the failing test was supposed to catch (and it did catch it!). But I think it makes sense for `concatenate` to descend through RecordArrays—the same motivation isn't there. (You can't overload `concatenate` the way you can overload ufuncs.)

Reply:
I can see that you didn't actually change `_numpy.py:215` in the branch, so I'm resolving this thread.